File: bot-api/python/src/robocode_tank_royale/bot_api/internal/intent_validator.py
import math
from typing import Any, Set, Optional
from ..bot_exception import BotException
from ..graphics import Color
from ..util.math_util import MathUtil
from ..constants import (
    MAX_SPEED,
    MAX_TURN_RATE,
    MAX_GUN_TURN_RATE,
    MAX_RADAR_TURN_RATE,
    DECELERATION,
    ACCELERATION,
    MAX_NUMBER_OF_TEAM_MESSAGES_PER_TURN,
    TEAM_MESSAGE_MAX_SIZE,
)


class IntentValidator:
    @staticmethod
    def validate_firepower(firepower: float) -> float:
        if math.isnan(firepower):
            raise ValueError("'firepower' cannot be NaN")
        return firepower

    @staticmethod
    def validate_turn_rate(turn_rate: float, max_turn_rate: float) -> float:
        if math.isnan(turn_rate):
            raise ValueError("'turn_rate' cannot be NaN")
        return MathUtil.clamp(turn_rate, -max_turn_rate, max_turn_rate)

    @staticmethod
    def validate_gun_turn_rate(gun_turn_rate: float, max_gun_turn_rate: float) -> float:
        if math.isnan(gun_turn_rate):
            raise ValueError("'gun_turn_rate' cannot be NaN")
        return MathUtil.clamp(gun_turn_rate, -max_gun_turn_rate, max_gun_turn_rate)

    @staticmethod
    def validate_radar_turn_rate(radar_turn_rate: float, max_radar_turn_rate: float) -> float:
        if math.isnan(radar_turn_rate):
            raise ValueError("'radar_turn_rate' cannot be NaN")
        return MathUtil.clamp(radar_turn_rate, -max_radar_turn_rate, max_radar_turn_rate)

    @staticmethod
    def validate_target_speed(target_speed: float, max_speed: float) -> float:
        if math.isnan(target_speed):
            raise ValueError("'target_speed' cannot be NaN")
        return MathUtil.clamp(target_speed, -max_speed, max_speed)

    @staticmethod
    def validate_max_speed(max_speed: float) -> float:
        return MathUtil.clamp(max_speed, 0, MAX_SPEED)

    @staticmethod
    def validate_max_turn_rate(max_turn_rate: float) -> float:
        return MathUtil.clamp(max_turn_rate, 0, MAX_TURN_RATE)

    @staticmethod
    def validate_max_gun_turn_rate(max_gun_turn_rate: float) -> float:
        return MathUtil.clamp(max_gun_turn_rate, 0, MAX_GUN_TURN_RATE)

    @staticmethod
    def validate_max_radar_turn_rate(max_radar_turn_rate: float) -> float:
        return MathUtil.clamp(max_radar_turn_rate, 0, MAX_RADAR_TURN_RATE)

    @staticmethod
    def get_new_target_speed(speed: float, distance: float, max_speed: float) -> float:
        if distance < 0:
            return -IntentValidator.get_new_target_speed(-speed, -distance, max_speed)

        target_speed: float
        if math.isinf(distance):
            target_speed = max_speed
        else:
            target_speed = min(max_speed, IntentValidator._get_max_speed_for_distance(distance))

        abs_deceleration = abs(DECELERATION)
        if speed >= 0:
            return MathUtil.clamp(target_speed, speed - abs_deceleration, speed + ACCELERATION)
        else:
            return MathUtil.clamp(
                target_speed,
                speed - ACCELERATION,
                speed + IntentValidator._get_max_deceleration(-speed),
            )

    @staticmethod
    def _get_max_speed_for_distance(distance: float) -> float:
        abs_deceleration = abs(DECELERATION)
        deceleration_time = max(
            1,
            math.ceil((math.sqrt((4 * 2 / abs_deceleration) * distance + 1) - 1) / 2),
        )
        if math.isinf(deceleration_time):
            return MAX_SPEED

        deceleration_distance = (
            (deceleration_time / 2) * (deceleration_time - 1) * abs_deceleration
        )
        return ((deceleration_time - 1) * abs_deceleration) + (
            (distance - deceleration_distance) / deceleration_time
        )

    @staticmethod
    def _get_max_deceleration(speed: float) -> float:
        abs_deceleration = abs(DECELERATION)
        deceleration_time = speed / abs_deceleration
        acceleration_time = 1 - deceleration_time
        return (
            min(1, deceleration_time) * abs_deceleration
            + max(0, acceleration_time) * ACCELERATION
        )

    @staticmethod
    def get_distance_traveled_until_stop(speed: float, max_speed: float) -> float:
        speed = math.fabs(speed)
        distance = 0.0
        while speed > 0:
            speed = IntentValidator.get_new_target_speed(speed, 0, max_speed)
            distance += speed
        return distance

    @staticmethod
    def validate_teammate_id(teammate_id: Optional[int], teammate_ids: Set[int]) -> None:
        if teammate_id is not None and teammate_id not in teammate_ids:
            raise ValueError("No teammate was found with the specified 'teammate_id'")

    @staticmethod
    def validate_team_message(message: Any, current_team_message_count: int) -> None:
        if current_team_message_count >= MAX_NUMBER_OF_TEAM_MESSAGES_PER_TURN:
            raise BotException(
                f"The maximum number team messages has already been reached: {MAX_NUMBER_OF_TEAM_MESSAGES_PER_TURN}"
            )
        if message is None:
            raise ValueError("The 'message' of a team message cannot be null")

    @staticmethod
    def validate_team_message_size(json_message_str: str) -> None:
        if len(json_message_str.encode("utf-8")) > TEAM_MESSAGE_MAX_SIZE:
            raise ValueError(
                f"The team message is larger than the limit of {TEAM_MESSAGE_MAX_SIZE} bytes (compact JSON format)"
            )

    @staticmethod
    def color_to_schema(color: Optional[Color]) -> Optional[str]:
        return color.to_color_schema() if color else None

File: bot-api/dotnet/api/src/internal/IntentValidator.cs
using System;
using System.Collections.Generic;
using Robocode.TankRoyale.BotApi;
using Robocode.TankRoyale.BotApi.Graphics;
using Robocode.TankRoyale.BotApi.Util;

namespace Robocode.TankRoyale.BotApi.Internal;

public static class IntentValidator
{
    public static double ValidateFirepower(double firepower)
    {
        if (double.IsNaN(firepower)) throw new ArgumentException("'firepower' cannot be NaN");
        return firepower;
    }

    public static double ValidateTurnRate(double turnRate, double maxTurnRate)
    {
        if (double.IsNaN(turnRate)) throw new ArgumentException("'TurnRate' cannot be NaN");
        return Math.Clamp(turnRate, -maxTurnRate, maxTurnRate);
    }

    public static double ValidateGunTurnRate(double gunTurnRate, double maxGunTurnRate)
    {
        if (double.IsNaN(gunTurnRate)) throw new ArgumentException("'GunTurnRate' cannot be NaN");
        return Math.Clamp(gunTurnRate, -maxGunTurnRate, maxGunTurnRate);
    }

    public static double ValidateRadarTurnRate(double radarTurnRate, double maxRadarTurnRate)
    {
        if (double.IsNaN(radarTurnRate)) throw new ArgumentException("'RadarTurnRate' cannot be NaN");
        return Math.Clamp(radarTurnRate, -maxRadarTurnRate, maxRadarTurnRate);
    }

    public static double ValidateTargetSpeed(double targetSpeed, double maxSpeed)
    {
        if (double.IsNaN(targetSpeed)) throw new ArgumentException("'TargetSpeed' cannot be NaN");
        return Math.Clamp(targetSpeed, -maxSpeed, maxSpeed);
    }

    public static double ValidateMaxSpeed(double maxSpeed)
    {
        if (double.IsNaN(maxSpeed)) throw new ArgumentException("'MaxSpeed' cannot be NaN");
        return Math.Clamp(maxSpeed, 0, Constants.MaxSpeed);
    }

    public static double ValidateMaxTurnRate(double maxTurnRate)
    {
        if (double.IsNaN(maxTurnRate)) throw new ArgumentException("'MaxTurnRate' cannot be NaN");
        return Math.Clamp(maxTurnRate, 0, Constants.MaxTurnRate);
    }

    public static double ValidateMaxGunTurnRate(double maxGunTurnRate)
    {
        if (double.IsNaN(maxGunTurnRate)) throw new ArgumentException("'MaxGunTurnRate' cannot be NaN");
        return Math.Clamp(maxGunTurnRate, 0, Constants.MaxGunTurnRate);
    }

    public static double ValidateMaxRadarTurnRate(double maxRadarTurnRate)
    {
        if (double.IsNaN(maxRadarTurnRate)) throw new ArgumentException("'MaxRadarTurnRate' cannot be NaN");
        return Math.Clamp(maxRadarTurnRate, 0, Constants.MaxRadarTurnRate);
    }

    public static double GetNewTargetSpeed(double speed, double distance, double maxSpeed)
    {
        if (distance < 0)
            return -GetNewTargetSpeed(-speed, -distance, maxSpeed);

        var targetSpeed = double.IsPositiveInfinity(distance) ? maxSpeed : Math.Min(GetMaxSpeed(distance), maxSpeed);

        double absDeceleration = Math.Abs(Constants.Deceleration);
        return speed >= 0
            ? Math.Clamp(targetSpeed, speed - absDeceleration, speed + Constants.Acceleration)
            : Math.Clamp(targetSpeed, speed - Constants.Acceleration, speed + GetMaxDeceleration(-speed));
    }

    private static double GetMaxSpeed(double distance)
    {
        double absDeceleration = Math.Abs(Constants.Deceleration);
        var decelerationTime =
            Math.Max(1, Math.Ceiling((Math.Sqrt((4 * 2 / absDeceleration) * distance + 1) - 1) / 2));
        if (double.IsPositiveInfinity(decelerationTime))
            return Constants.MaxSpeed;

        var decelerationDistance = (decelerationTime / 2) * (decelerationTime - 1) * absDeceleration;
        return ((decelerationTime - 1) * absDeceleration) + ((distance - decelerationDistance) / decelerationTime);
    }

    private static double GetMaxDeceleration(double speed)
    {
        double absDeceleration = Math.Abs(Constants.Deceleration);
        var decelerationTime = speed / absDeceleration;
        var accelerationTime = 1 - decelerationTime;

        return Math.Min(1, decelerationTime) * absDeceleration +
               Math.Max(0, accelerationTime) * Constants.Acceleration;
    }

    public static double GetDistanceTraveledUntilStop(double speed, double maxSpeed)
    {
        speed = Math.Abs(speed);
        double distance = 0;
        while (speed > 0)
            distance += (speed = GetNewTargetSpeed(speed, 0, maxSpeed));

        return distance;
    }

    public static void ValidateTeammateId(int? teammateId, ICollection<int> teammateIds)
    {
        if (teammateId != null && !teammateIds.Contains((int)teammateId))
        {
            throw new ArgumentException("No teammate was found with the specified 'teammateId': " + teammateId);
        }
    }

    public static void ValidateTeamMessage(object message, int currentTeamMessageCount)
    {
        if (currentTeamMessageCount == IBaseBot.MaxNumberOfTeamMessagesPerTurn)
            throw new InvalidOperationException(
                "The maximum number team massages has already been reached: " +
                IBaseBot.MaxNumberOfTeamMessagesPerTurn);

        if (message == null)
            throw new ArgumentException("The 'message' of a team message cannot be null");
    }

    public static void ValidateTeamMessageSize(string json)
    {
        var bytes = System.Text.Encoding.UTF8.GetBytes(json);
        if (bytes.Length > IBaseBot.TeamMessageMaxSize)
            throw new ArgumentException(
                $"The team message is larger than the limit of {IBaseBot.TeamMessageMaxSize} bytes (compact JSON format)");
    }

    public static string ColorToHex(Color? color) => color == null ? null : "#" + ColorUtil.ToHex(color);
}
